packages/client/src/runtime/core/types/Extensions.ts

```diff
@@ -21,30 +25,178 @@ export type Args = InternalArgs
+  GetFindResult<TypeMap['model'][M]['findFirstOrThrow']['payload'], { select: S }>
 
-export type GetMaybeITXClient<Base extends Record<any, any>, C extends Args['client']> =
-  MakeITXPropsOptional<GetClient<Base, C>>
+/** Model */
 
-type MakeITXPropsOptional<C> = Omit<C, ITXClientDenyList> & {
-  [K in Extract<keyof C, ITXClientDenyList>]?: C[K]
+export type DynamicModelExtensionArgs<M_, TypeMap extends Record<string, any>, ExtArgs extends Record<string, any>> = {
+  [K in keyof M_]:
+    K extends '$allModels'
+    ? & { [P in keyof M_[K]]?: unknown }
+      & { [K: symbol]: {} }
+    : K extends TypeMap['meta']['modelProps']
+      ? & { [P in keyof M_[K]]?: unknown }
+        & { [K: symbol]: { ctx: DynamicModelExtensionThis<TypeMap, ModelKey<TypeMap, K>, ExtArgs> & { name: ModelKey<TypeMap, K> } } }
+      : never
 }
 
-export type ReadonlySelector<T> = T extends unknown ? {
-  readonly [K in keyof T as K extends 'include' | 'select' ? K : never]: ReadonlyDeep<T[K]>
+type DynamicModelExtensionThis<TypeMap extends Record<string, any>, M extends keyof any, ExtArgs extends Record<string, any>> = {
+  [P in keyof TypeMap['model'][M] | keyof ExtArgs['model'][Uncapitalize<M & string>]]:
+    P extends Operation
+    ? HasAllOptionalKeys<TypeMap['model'][M][P]['args']> extends 1
+      ? <A extends TypeMap['model'][M][P]['args']>(args?: A) =>
+          PrismaPromise<GetOperationResult<TypeMap['model'][M][P]['payload'], A, P>>
+      : <A extends TypeMap['model'][M][P]['args']>(args: A) =>
+          PrismaPromise<GetOperationResult<TypeMap['model'][M][P]['payload'], A, P>>
+    : Return<ExtArgs['model'][Uncapitalize<M & string>][P]>
 } & {
-  [K in keyof T as K extends 'include' | 'select' ? never : K]: T[K]
-} : never
+  [K: symbol]: { types: TypeMap['model'][M] }
+}
+
+/** Client */
+
+export type DynamicClientExtensionArgs<C_, TypeMap extends Record<string, any>, TypeMapCb extends TypeMapCbDef, ExtArgs extends Record<string, any>> = {
+  [P in keyof C_]: unknown
+} & {
+  [K: symbol]: { ctx: Optional<DynamicClientExtensionThis<TypeMap, TypeMapCb, ExtArgs>, ITXClientDenyList> }
+}
+
+export type DynamicClientExtensionThis<TypeMap extends Record<any, any>, TypeMapCb extends TypeMapCbDef, ExtArgs extends Record<string, any>> = {
+  [P in keyof TypeMap['other'] | keyof TypeMap['model'] | keyof ExtArgs['client'] as Uncapitalize<P & string>]:
+    P extends Operation
+    ? <A extends TypeMap['other'][P]['args']>(...args: A extends any[] ? A : [A]) =>
+      PrismaPromise<GetOperationResult<TypeMap['other'][P]['payload'], A, P>>
+    : P extends keyof TypeMap['model']
+      ? DynamicModelExtensionThis<TypeMap, P, ExtArgs>
+      : Return<ExtArgs['client'][P]>
+} & {
+  $extends: ExtendsHook<'extends', TypeMapCb, ExtArgs>
+  $transaction<R>(fn: (client: Omit<DynamicClientExtensionThis<TypeMap, TypeMapCb, ExtArgs>, ITXClientDenyList>, options?: { maxWait?: number, timeout?: number, isolationLevel?: string }) => Promise<R>): Promise<R>
```

Comment:
This loses strong typing for `isolationLevel` on extended client and makes it available where it should not (mongo).
I guess we should put `isolationLevel` into the type map as well?

Reply:
Good catch, thanks.